File: LowLevelDesign/SOLID_Assignment_1/C#/Rules/ConcreteRules/MaxAmountRule.cs
using Rules;
namespace Rules.ConcreteRules
{
    public class MaxAmountRule : IExpenseRule
    {
        private readonly double _maxAmount;

        public MaxAmountRule(double maxAmount)
        {
            this._maxAmount = maxAmount;
        }

        public Violation? ValidateExpense(Models.Expense expense)
        {
            if (expense.Amount > this._maxAmount)
            {
                return Violation.CreateViolation($"Expense with Id {expense.GetExpenseId()} has amount {expense.GetAmount()} exceeds the maximum allowed amount of {_maxAmount}.");
            }
            return null;
        }
    }
}

File: LowLevelDesign/SOLID_Assignment_1/lib/Rules/Implementations/MaxAmountRule.java
package LowLevelDesign.SOLID_Assignment_1.lib.Rules.Implementations;

import java.util.Optional;

import LowLevelDesign.SOLID_Assignment_1.lib.Models.Expense;
import LowLevelDesign.SOLID_Assignment_1.lib.Rules.IExpenseRule;
import LowLevelDesign.SOLID_Assignment_1.lib.Rules.Violation;

public class MaxAmountRule implements IExpenseRule {

    private final double maxAmount;

    public MaxAmountRule(double maxAmount) {
        this.maxAmount = maxAmount;
    }

    @Override
    public Optional<Violation> check(Expense expense) {
        if (expense.getAmount() > maxAmount) {
            return Optional.of(Violation
                    .of("Expense amount " + expense.getAmount() + " exceeds maximum allowed amount of " + maxAmount));
        } else {
            return Optional.empty();
        }
    }

}
